pkg/apis/nfd/validate/validate.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	k8sQuantity "k8s.io/apimachinery/pkg/api/resource"
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+)
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+	// Default error message for invalid label/annotation keys (untyped string)
+	ErrUnprefixedKeysNotAllowed = fmt.Errorf("namespace without prefix/ not allowed")
+	// Default error for invalid taint effect
+	ErrInvalidTaintEffect = fmt.Errorf("invalid taint effect")
+	// Default error for empty taint effect
+	ErrEmptyTaintEffect = fmt.Errorf("empty taint effect")
+)
+
+// Label validates a label key and value and returns an error if the key or
+// value is invalid.
+func Label(key, value string) error {
+	//Validate label key and value
+	if err := k8svalidation.IsQualifiedName(key); len(err) > 0 {
+		return fmt.Errorf("invalid label key %q: %s", key, strings.Join(err, "; "))
+	}
+	// Check label namespace, filter out if ns is not whitelisted
+	ns, _ := splitNs(key)
+	// And is not empty
+	if ns == "" {
+		return ErrUnprefixedKeysNotAllowed
+	}
+	if ns != nfdv1alpha1.FeatureLabelNs && ns != nfdv1alpha1.ProfileLabelNs &&
+		!strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) && !strings.HasSuffix(ns, nfdv1alpha1.ProfileLabelSubNsSuffix) {
+		return ErrNSNotAllowed
```

Comment:
Hmm, I think we should deny `kubernetes.io` here but allow others, as that's the default setup. Similar to what we do for annotations and ERs and taints below.

```suggestion
		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
			return ErrNSNotAllowed
		}
```